### PR TITLE
Refactor Babylon->GLTF2 exporter to use shared project

### PR DESCRIPTION
**Pull request by Drigax**
This PR removes the engine-specific dependencies from the 3dsMax Babylon->GLTF2 exporter, and moves it to a shared project, Babylon2GLTF intended to be consumed from the 3dsMax and Maya exporters.

### CODE REVIEW

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs [hunk 1]
```diff
@@ -35,7 +77,7 @@ internal interface IGLTFMaterialExporter : IMaterialExporter
         /// <param name="raiseWarningAction">Callback function to raise warnings.</param>
         /// <param name="raiseErrorAction">Callback function to raise errors.</param>
         /// <returns>The exported GLTF material.</returns>
-        GLTFMaterial ExportGLTFMaterial(BabylonExporter exporter, GLTF gltf, IIGameMaterial material, Func<string, string, string> tryWriteImageFunc,
+        GLTFMaterial ExportGLTFMaterial(ExportParameters exportParameters, GLTF gltf, IIGameMaterial material, Func<string, string, string> tryWriteImageFunc,
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs]: @elpie89 , i wanted to change this interface, i'm not sure if you'd really need the BabylonExporter as a whole to export the MaxMaterial correctly, but since this is a breaking change, let me know if we should have an alternative.

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs [hunk 2]
```diff
@@ -36,33 +38,35 @@ internal partial class BabylonExporter
         public static string exporterVersion = "1.4.2";
         public float scaleFactor = 1.0f;
 
-        void ReportProgressChanged(int progress)
+        public const int MaxSceneTicksPerSecond = 4800; //https://knowledge.autodesk.com/search-result/caas/CloudHelp/cloudhelp/2016/ENU/MAXScript-Help/files/GUID-141213A1-B5A8-457B-8838-E602022C8798-htm.html
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs]: Right now, i put this constant here, but it may be prudent to do another minor refactor to move some of the magic numbers floating around the codebase to a more centralized place with links to documentation for clarity

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs [hunk 3]
```diff
@@ -10,10 +11,11 @@
 using System.Text;
 using System.Windows.Forms;
 using Color = System.Drawing.Color;
+using Utilities;
 
 namespace Max2Babylon
 {
-    internal partial class BabylonExporter
+    internal partial class BabylonExporter : ILoggingProvider
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs]: This is probably one of the ugliest parts of this refactor, I'm not really a huge fan of adding this interface, but it was the easiest way to reconcile all the areas where we're logging in the GLTF converter. Maybe instead have a logging class that handles writing to the UI implementation?

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs [hunk 4]
```diff
@@ -444,15 +438,15 @@ public void Export(ExportParameters exportParameters)
                 {
                     ExportSkin(skin, babylonScene);
                 }
-            }
+            }
+
+            // Animation group
+            RaiseMessage("Export animation groups");
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs]: Not sure why we have two paths for exporting animation groups between babylon/gltf. My solution was to just have the gltf exporter use the babylon scene's already exported animation groups.

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs [hunk 5]
```diff
@@ -79,6 +79,8 @@ private void ExporterForm_Load(object sender, EventArgs e)
             Tools.PrepareCheckBox(chkKHRTextureTransform, Loader.Core.RootNode, "babylonjs_khrTextureTransform");
             Tools.PrepareCheckBox(chkKHRMaterialsUnlit, Loader.Core.RootNode, "babylonjs_khr_materials_unlit");
             Tools.PrepareCheckBox(chkExportMaterials, Loader.Core.RootNode, "babylonjs_export_materials", 1);
+            Tools.PrepareCheckBox(chkExportMorphTangents, Loader.Core.RootNode, "babylonjs_export_Morph_Tangents", 0);
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: Currently we also have a checkbox for exporting morph target tangents and normals in the other exporter form @PatrickRyanMS , would it be alright to move them to the main exporter dialog?

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs [hunk 6]
```diff
@@ -196,7 +201,8 @@ private async Task<bool> DoExport(ExportItem exportItem, bool multiExport = fals
             exporter = new BabylonExporter();
             if (!string.IsNullOrWhiteSpace(txtTextureName.Text))
             {
-                exporter.relativeTextureFolder = Tools.GetPathRelativeToModel(Tools.UnformatPath(txtTextureName.Text), Tools.UnformatPath(txtModelName.Text));
+                // TODO:
```
Comment (Drigax) [inline 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: Oops, I need to re-add this for the next iteration.

File: SharedProjects/Babylon2GLTF/GLTFExporter.Animation.cs [hunk 7]
```diff
@@ -308,28 +226,10 @@ private void ExportBoneAnimation(GLTFAnimation gltfAnimation, int startFrame, in
 
             if (babylonBone.animation != null && babylonBone.animation.property == "_matrix")
             {
-                RaiseMessage("GLTFExporter.Animation | Export animation of bone named: " + babylonBone.name, 2);
+                logger.RaiseMessage("GLTFExporter.Animation | Export animation of bone named: " + babylonBone.name, 2);
 
                 var babylonAnimation = babylonBone.animation;
 
-                // Optimize animation
```
Comment (Drigax) [inline SharedProjects/Babylon2GLTF/GLTFExporter.Animation.cs]: I removed this optimization pass because we should already do it when exporting babylon. we should minimize our changes to the input Babylon file and just convert what we have to GLTF

File: SharedProjects/Babylon2GLTF/GLTFExporter.Animation.cs [hunk 8]
```diff
@@ -465,7 +365,7 @@ private GLTFAccessor _createAndPopulateInput(GLTF gltf, BabylonAnimation babylon
                 numKeys++;
                 float inputValue = babylonAnimationKey.frame;
                 if (offsetToStartAtFrameZero) inputValue -= startFrame;
-                inputValue /= Loader.Global.FrameRate;
+                inputValue /= babylonAnimation.framePerSecond;
```
Comment (Drigax) [inline SharedProjects/Babylon2GLTF/GLTFExporter.Animation.cs]: this may cause a regression, but it may also be a bug fix, @PatrickRyanMS it may be good to have a test where we have an animation authored with a different frame rate than scene, (if that's possible in both programs)

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs (hunk 1 above)
Comment (elpie89) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs]: @Selmar worked on this interface from the beginning, I didn't touch the IMaterialExporter...he should know better than me
What do you think @Selmar ?

File: 3ds Max/Max2Babylon/Exporter/GlobalVertex.cs [hunk 9]
```diff
@@ -1,4 +1,5 @@
 ﻿using Autodesk.Max;
+using Utilities;
```
Comment (sebavan) [inline 3ds Max/Max2Babylon/Exporter/GlobalVertex.cs]: Why is this needed if no code changed in the file ?

**Comment by sebavan** [comment]
Understood better the change for the designer :-) Looks good, be just carefull to keep in the latest changes and not lose some with the merge, designer.cs are merging really poorly :-(

File: 3ds Max/Max2Babylon/Exporter/GlobalVertex.cs (hunk 9 above)
Comment (Drigax) [reply 3ds Max/Max2Babylon/Exporter/GlobalVertex.cs]: Not needed. I'll remove this statement

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs [hunk 10]
```diff
@@ -1,31 +1,73 @@
...
 using BabylonExport.Entities;
-using GLTFExport.Entities;
-
+using GLTFExport.Entities;
+using Utilities;
+
 namespace Max2Babylon
 {
-    
+    internal class MaxGLTFMaterialExporter : IGLTFMaterialExporter
```
Comment (sebavan) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs]: Should not this be moved in a separate file ? 

Sounds weird in a IMaterialExporter one considering I is for interface ?

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.Material.cs [hunk 11]
```diff
@@ -3,6 +3,7 @@
 using System.Collections.Generic;
 using Autodesk.Max;
 using BabylonExport.Entities;
+using Utilities;
```
Comment (sebavan) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.Material.cs]: as previously not sure it is necessary here

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs (hunk 3 above)
Comment (sebavan) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.cs]: yup you could have a logger implementing the interface in both max and maya and then you could pass the interface in the exporter constructors ???

File: 3ds Max/Max2Babylon/MaxScriptManager.cs [hunk 12]
```diff
@@ -76,8 +78,8 @@ public static ExportParameters InitParameters(string outputPath)
             exportParameters.autoSave3dsMaxFile = Loader.Core.RootNode.GetBoolProperty("babylonjs_autosave");
             exportParameters.exportOnlySelected = Loader.Core.RootNode.GetBoolProperty("babylonjs_onlySelected");
             exportParameters.exportTangents = Loader.Core.RootNode.GetBoolProperty("babylonjs_exporttangents");
-            exportParameters.scaleFactor = Loader.Core.RootNode.GetStringProperty("babylonjs_txtScaleFactor", "1");
-            exportParameters.txtQuality = Loader.Core.RootNode.GetStringProperty("babylonjs_txtCompression", "100");
+            exportParameters.scaleFactor = float.Parse(Loader.Core.RootNode.GetStringProperty("babylonjs_txtScaleFactor", "1"));
```
Comment (sebavan) [inline 3ds Max/Max2Babylon/MaxScriptManager.cs]: Should that be TryParse in case not float values are in to prevent uncaught exceptions ?

(valid for the next line as well)

File: SharedProjects/Utilities/TextureUtilities.cs [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+using System.Collections.Generic;
+using System.Drawing;
+using System.Drawing.Imaging;
+using System.IO;
+using BabylonExport.Entities;
+using GLTFExport.Entities;
+
+namespace Utilities
+{
+    class TextureUtilities
```
Comment (sebavan) [inline SharedProjects/Utilities/TextureUtilities.cs]: could the class be static as well ? reenforcing the intent of utils contained in its name ?

**Comment by sebavan** [comment]
Just added a few #neat comments and it would be good to go I guess

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.Material.cs (hunk 11 above)
Comment (Drigax) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.Material.cs]: Yes, we include it for float[] extension methods like float[].AlmostEqualTo

File: SharedProjects/Utilities/TextureUtilities.cs (hunk 13 above)
Comment (Drigax) [reply SharedProjects/Utilities/TextureUtilities.cs]: good point, I'll add that as well.

File: 3ds Max/Max2Babylon/MaxScriptManager.cs (hunk 12 above)
Comment (Drigax) [reply 3ds Max/Max2Babylon/MaxScriptManager.cs]: Yes, we should make sure we don't throw exceptions. I'll change that to TryParse

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs (hunk 10 above)
Comment (Drigax) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs]: makes sense, I'll move the implementation to a separate file.

**Comment by Drigax** [comment]
the designer.cs file for the export form was a painstaking hand merge, but it looks like everything that was merged in was preserved there. I'm pretty sure that the GLTF exporter doesn't include the Maya related changes, but I'll double check before merging.

File: 3ds Max/Max2Babylon/Forms/ExporterForm.cs (hunk 6 above)
Comment (Drigax) [reply 3ds Max/Max2Babylon/Forms/ExporterForm.cs]: readded, this definitely exposed some additional spaghetti in our path related code. did some additional refactoring to remove duplicate functionality (we had two pretty similar functions for relative path formatting in Tools.cs) and moved the 3d software independent code to a Utilities class.

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs (hunk 1 above)
Comment (Selmar) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs]: Looking at the current code, the only thing we need are the export parameters. This should be fine.

I haven't suggested this before, but I feel the material exportation in the babylon exporter could make use of this same interface, as opposed to having all the exports in one big function as it is now. It should not be very difficult, just need to know the ClassID, implement the interface and move the code, although some parameters may need to be added/changed (I haven't looked at this since I made it).

Comment (Selmar) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs]: Of course, as it isn't used by anyone, the IBabylonMaterialExporter is currently not complete in any way.

**Comment by sebavan** [comment]
CI report is all good https://dev.azure.com/babylonjs/ContinousIntegration/_build/results?buildId=1042

Just the API seems to have got stuck.

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs (hunk 1 above)
Comment (Drigax) [reply 3ds Max/Max2Babylon/Exporter/BabylonExporter.IMaterialExporter.cs]: Sounds good, thanks for the input 👍

**Comment by Drigax** [comment]
Maya Material export fixes are in. :) checking in after CI finishes. did some rudimentary testing with relatively simple scenes, but so far I haven't seen any regression in basic functionality.
